### ISSUE TITLE
Characters get inserted with wrong order, abbreviation characters don't get removed and few are missing

### ISSUE THREAD

**Issue by novoid** [autokey/autokey#763]
### Has this issue already been reported?

- [X] I have searched through the existing issues.

### Is this a question rather than an issue?

- [X] This is not a question.

### What type of issue is this?

Bug

### Which Linux distribution did you use?

Xubuntu 22.04 LTS

### Which AutoKey GUI did you use?

GTK

### Which AutoKey version did you use?

0.95.10

### How did you install AutoKey?

Ubuntu + apt

### Can you briefly describe the issue?

After upgrading from Xubuntu 20.04 LTS to Xubuntu 22.04 LTS, AutoKey is inserting snippets from Scripts and Phrases with wrong order of characters and sometimes, a few characters are missing. Also: abbreviation characters are not removed properly.

Example: `$$dd` (my abbreviation) is inserting `$203-01-25` which should be `2023-01-25`.

Script definition:
```
output = system.exec_command("date '+%Y-%m-%d'")
keyboard.send_keys(output)
```

Here are a few unfiltered invocations:
```
$2023-01-25
$2023-01-25
$$2023-01-25
$2023-01-25
$2023-0125
2023-01-25
$203-01-25
2023-01-25
```

Two out of eight times, the script worked as expected. I also had multiple incidents where I got something like `202-0125` (missing characters).

I saw multiple statements that there is an issue with shortcuts and Wayland. I don't have an issue with the abbreviations (I'm not using shortcut hotkeys), only with inserted characters. Maybe you should specify the issues related to Wayland in the Readme a bit more detailed if this might be also related to Wayland.

Phrases do have the option of using the clipboard as discussed on #380 - but this is a rather dirty workaround IMHO and is not offered for scripts like my "today's date" script example above.

My brother has the same issue after upgrading to Xubuntu 22.04 LTS. So it's not just my computer that is affected.

### Can the issue be reproduced?

Always

### What are the steps to reproduce the issue?

1. Define a script (or phrase)
2. Assign an abbreviation string
3. Invoke the abbreviation string

### What should have happened?

The expansion should have replaced the abbreviation string with the resulting string.

### What actually happened?

Abbreviation characters did not get removed, characters of resulting strings got switched or removed.  

### Do you have screenshots?

I don't think a screenshot is helping here.

**Comment by Elliria** [comment]
Does this happen when pasting into any program or in certain ones? There are known programs (like Firefox and LibreOffice) that behave in this way and often adding a ```time.sleep()``` delay into a script in one or more places can solve it. Such a delay can also sometimes be necessary when running system commands, as you're doing in your script.

You could try this, adjusting the interval if you need to, to see if it will make the script behave the way you'd like it to:
```python
output = system.exec_command("date '+%Y-%m-%d'")
time.sleep(0.1)
keyboard.send_keys(output)
```

**Comment by novoid** [comment]
> Does this happen when pasting into any program or in certain ones? There are known programs (like Firefox and LibreOffice) that behave in this way and often adding a `time.sleep()` delay into a script in one or more places can solve it. Such a delay can also sometimes be necessary when running system commands, as you're doing in your script.

I'm using this snippet mostly in Firefox and GNU/Emacs.

> You could try this, adjusting the interval if you need to, to see if it will make the script behave the way you'd like it to:
> 
> ```python
> output = system.exec_command("date '+%Y-%m-%d'")
> time.sleep(0.1)
> keyboard.send_keys(output)
> ```

Here are seven samples from GNU/Emacs 28.2 using your sleep instruction right now:
```
2-10-32025
02-01-2532
0252-10-32
023-01-252
023-01-252
02-01-2532
02-01-2532
```

And again in my Firefox:
```
$2023-01-25
2023-01-25
2023-01-25
2023-01-25
2023-01-25
2023-01-25
$2023-01-25
2023-01-25
```

So in Firefox, this issue seems to be a bit reduced. In GNU/Emacs, the issue renders AutoKey useless. And the nature of the output is different with both.

Same result with a higher value of 0.3.

However, as I'd think, the delay should better be in between inserting two characters and not in between generating the output string and pushing the string to the keyboard buffer, right? 

My zsh in xfce4-terminal v0.8.10 does not seem to have this issue: strings are inserted without an issue. Maybe it's an issue for XOrg/Wayland apps and not for CLI apps? 

Now it gets weird: `emacs -q -nw` (GNU/Emacs with default settings in terminal with not Xorg/Wayland) does not show the issue. Then I started `emacs -q` (GNU/Emacs with default setting; again omitting my personal configuration files) is also working fine. So at least with GNU/Emacs, I do have this weird behavior that the issue arises only when invoked with [my personal configuration](https://github.com/novoid/dot-emacs/blob/master/config.org).

If I could not reproduce the issue in Firefox, I'd say I may have an GNU/Emacs issue.

**Comment by Elliria** [comment]
The good news is that you're able to use the script successfully in some situations and I think you solved it the moment you mentioned **Xorg/Wayland.** AutoKey is a **Xorg** application and will not function in a **Wayland** session. Have you tried using a **Xorg** session and does the script work fine when you do?

Also, although it turns out that this probably has nothing to do with your issue, I just thought I'd mention that sometimes a significant delay (sometimes as much as several seconds) is necessary to solve timing problems and I can't say it would surprise me to hear of a situation in which it was necessary to put a delay between each character that's printed, but I don't think anyone has needed to go that far yet.

**Comment by josephj11** [comment]
Welcome to the AutoKey community, @novoid !

One of our top priorities is to get #600 implemented and released. It doesn't fix anything, but it will warn people not to use Wayland if they want to use AutoKey.

That aside, it is quite easy to rewrite your script to place output into the clipboard and then emit a Ctrl+V to paste it.

We have had a huge number of issues exactly like this one. Most are solved by pasting from the clipboard. See [this](https://github.com/autokey/autokey/wiki/Emitting-Keyboard-Events) for more details.

**Comment by realalexandergeorgiev** [comment]
Having similar issues on X server. It would be great to have a delay between each keypress event. Using the KeepassXC paste method it always works, but it has something like 100ms delay between each key event.

**Comment by Elliria** [comment]
This is a demonstration based on @josephj11's [Function to Type Text Slowly](https://github.com/autokey/autokey/wiki/Advanced-Scripts#function-to-type-text-slowly) script on the **Advanced Scripts** page.
```python
# Define a function to type a string, character by character, with a specified delay between characters:
def type_slow(string, delay=0):
  # If the delay is less than zero, make it zero:
  if delay < 0:
    delay = 0

  # Type each character followed by the specified delay:
  for c in string:
    keyboard.send_keys(c)
    time.sleep(delay)

# Use the function with and without various delays:
type_slow("foo") # no delay
type_slow("foo",0.1) # a delay of 0.1 seconds
type_slow("foo",1) # a delay of 1 second
type_slow("foo",2) # a delay of 2 seconds
```

**Comment by novoid** [comment]
> The good news is that you're able to use the script successfully in some situations and I think you solved it the moment you mentioned **Xorg/Wayland.** AutoKey is a **Xorg** application and will not function in a **Wayland** session. Have you tried using a **Xorg** session and does the script work fine when you do?

No. As a matter of fact, I don't know how to switch to Xorg as I'm using the Xubuntu defaults and no login manager which would give me the ability to switch.

Independent of my issue, I'd like to propose an extension of the README comment about Wayland with a paragraph what's not working with Wayland. All known issues so far. Because as I was looking through the issues here, I got the impression that only the hotkeys aren't working because of Wayland. If Wayland is affecting more or less the whole functionality of AutoKey, you should add a statement accordingly so that any issue reported by Wayland-users is avoided in the first place.

Just my 2 cents.

**Comment by novoid** [comment]
Thank you for that snippet. I really need to learn more about the script language as it seems to be very powerful.

One question: do I need to define the `type_slow()` function in every script I'm using or can this be done once for all scripts I'm using in AutoKey?

```
def type_slow(string, delay=0):
  if delay < 0:
    # I'd like to live in an IT word where a negative number here would be handled as expected ;-) -> time travelling made easy.
    delay = 0
  for c in string:
    keyboard.send_key(c)
    time.sleep(delay)
output = system.exec_command("date '+%Y-%m-%d'")
type_slow(output, 0.01)
```

Even with this very small delay, my `$$dd` script is working fairly well:
```
2023-01-26
2023-01-26
2023-01-26
2023-01-26
2023-01-26
$023-01-26
$023-01-26
2023-01-26
2023-01-26
```

I may settle for a higher value that doesn't delay me too much. However, a noticeable delay feels totally cool to me because it reminds me of those (usually older) movies where computers type something on the screen ... ;-)

So for the moment, I seem to be able to continue working with AutoKey despite using Wayland (which was not my deliberate choice in the first place but Ubuntu's choice, btw).

Thank you for your help. I consider this issue solved using the workaround using the handy `type_slow()` function.

**Comment by novoid** [comment]
Okay, I do have one additional question:

```
def type_slow(string, delay=0):
  # If the delay is less than zero, make it zero:
  if delay < 0:
    delay = 0

  # Type each character followed by the specified delay:
  for c in string:
    keyboard.send_key(c)
    time.sleep(delay)

output = "One: Two\nThree"
type_slow(output, 0.02)
```

... this always result in:
```
one; two
three
```

Notice the lower case as well as the character switch from colon to semi-colon.

Is this a bug I should create a ticket for or some kind of desired behavior? In case of the latter, the workaround proposed isn't working for most of my snippets. :-(

Update: the first line is from the script definition and the second is the result after executing the script:
```
ABC!@#$%^&*()+-_<>/?,.[]{}-_
abc12345678()=--,.//,.[][]--
```

There are multiple characters that get replaced. It seems to me that the shift key is not pressed in the "virtual" keyboard.

**Update 2023-01-27: `keyboard.send_key(c)` -> `keyboard.send_keys(c)` fixed the issue.**

**Comment by Elliria** [comment]
> No. As a matter of fact, I don't know how to switch to Xorg as I'm using the Xubuntu defaults and no login manager which would give me the ability to switch.

You'll need to [disable "Automatic Login"](https://www.google.com/search?channel=fs&client=ubuntu&q=xubuntu+disable+automatic+login) and then switch to **Xorg**. You can re-enable "Automatic Login" afterwards.
 
> Independent of my issue, I'd like to propose an extension of the README comment about Wayland with a paragraph what's not working with Wayland.

We agree that this needs to be brought to the attention of users. A notice has been added as the [Important - AutoKey is not compatible with Wayland](https://github.com/autokey/autokey#important---autokey-is-not-compatible-with-wayland) section in the **README.rst** that has its contents displayed on the [AutoKey main page](https://github.com/autokey/autokey). We're also working on an update to the issue template so that the very first question will let users know that and ask which session type they're using.

**Comment by Elliria** [comment]
> Thank you for that snippet. I really need to learn more about the script language as it seems to be very powerful.

The [AutoKey wiki](https://github.com/autokey/autokey/wiki) has a couple of pages that will probably interest you:

- There's the [Scripting](https://github.com/autokey/autokey/wiki/Scripting) page that provides a quick introduction and summary of scripting in AutoKey.
- There's also the [Example Scripts](https://github.com/autokey/autokey/wiki/Example-Scripts) section with lots of examples you can check out and hopefully add to in time.
 

> One question: do I need to define the `type_slow()` function in every script I'm using or can this be done once for all scripts I'm using in AutoKey?

You can [run an AutoKey script from an AutoKey script](https://github.com/autokey/autokey/wiki/Advanced-Scripts#run-an-autokey-script-from-an-autokey-script), so you can put the Type Slow function into one script and use it from any of your other scripts.

> Even with this very small delay, my `$$dd` script is working fairly well:

I'm glad, and perhaps it will work flawlessly with a slight increase.

> However, a noticeable delay feels totally cool to me because it reminds me of those (usually older) movies where computers type something on the screen ... ;-)

Yeah, I get the same feeling from it.
 
> So for the moment, I seem to be able to continue working with AutoKey despite using Wayland (which was not my deliberate choice in the first place but Ubuntu's choice, btw).
> 
> Thank you for your help. I consider this issue solved using the workaround using the handy `type_slow()` function.

I'm glad it's working now.

**Comment by Elliria** [comment]
> Notice the lower case as well as the character switch from colon to semi-colon.
> 
> Is this a bug I should create a ticket for or some kind of desired behavior? In case of the latter, the workaround proposed isn't working for most of my snippets. :-(

It's not a bug. It was my oversight.
 
> There are multiple characters that get replaced. It seems to me that the shift key is not pressed in the "virtual" keyboard.

You're absolutely right. I should have noticed that in the script I provided (and I've fixed it in the snippet I provided above now). The AutoKey API provides two calls for sending characters:
* the singular ```keyboard.send_key()``` call
* the plural ```keyboard.send_keys()``` call

The singular call always forces lower-case characters and only allows you to send one character. The plural call accepts lower-case and upper-case characters and allows you to send as many characters as you like.

If you replace the singular call in your snippet with the plural call, the issue should be solved.

**Comment by Elliria** [comment]
Hey there, @josephj11, those scripts mentioned above should be updated. Would you like me to do it or would you like to? I'm hesitating to do it because I'm not sure of the purpose of the namespace in the one script, so I don't want to just get rid of it. I'm also not sure whether you have a preference for one syntax over another for the two different functions or would like to keep them both.

**Comment by josephj11** [comment]
> Having similar issues on X server. It would be great to have a delay between each keypress event. Using the KeepassXC paste method it always works, but it has something like 100ms delay between each key event.

I'm a big fan of KeepassXC. Its autotype feature is great. I think it allows you to configure the delays it uses, but the defaults usually work.

**Comment by josephj11** [comment]
@novoid AutoKey scripts are coded in native Python3, so they can do almost anything, limited only by your proficiency in Python and by your imagination! This also means that you can learn how to do things from the extensive documentation, tutorials, and Python examples all over the web.

In addition to @Elliria 's suggestion of using the AutoKey script engine to run common code, there are a couple of other ways to address this that you can read about in the comments in #248.

**Comment by josephj11** [comment]
@novoid Not just in old movies. I just watched Counterpart (which is around a year old) and "Management" uses these strange portable terminals that display what looks like a Linux CLI scrolling at old TTY speeds. If you like Sci Fi and haven't seen J. K. Simmons (the lead) before, check it out. He's an amazing actor.

**Comment by josephj11** [comment]
@Elliria I improved the new function and added the enhanced version.

The old one is in the Python2 branch of AutoKey and probably should be ignored.

i don't think anything there should get updated. Maybe we can delete the whole branch some day (after copying my problem reporting guide to the current repo).

**Comment by Elliria** [comment]
> In addition to @Elliria 's suggestion of using the AutoKey script engine to run common code, there are a couple of other ways to address this that you can read about in the comments in #248.

This might also be a good place to plug the new [Modules](https://github.com/autokey/autokey/wiki/Modules) page in the AutoKey wiki.

**Comment by Elliria** [comment]
> @Elliria I improved the new function and added the enhanced version.

I see that. Thanks for doing that. I'd like to make one additional suggestion that just occurred to me. Often, when we see these types of issues, the first character also gets mangled, so perhaps putting the delay before the character in the loop would take care of that, so instead of sending a character followed by a delay, you'd be sending a delay followed by a character. I just tried it and it works nicely here.
 
> The old one is in the Python2 branch of AutoKey and probably should be ignored.

I didn't realize that. I should pay more attention to the URLs. It's now removed from he post I made above, to avoid sending anyone there.
 
> i don't think anything there should get updated.

Agreed.

> Maybe we can delete the whole branch some day (after copying my problem reporting guide to the current repo).

Sure, but definitely not yet. We should harvest it for anything (including your problem-reporting guide) that's not in the current wiki first.
